### PR TITLE
[Kanal D]Added support for new site

### PR DESCRIPTION
**Pull request by kosantosbik**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Added support for Kanal D TV channel's website.

### CODE REVIEW

File: youtube_dl/extractor/kanald.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+from __future__ import unicode_literals
+
+from re import findall
+
+from .common import InfoExtractor
+from ..compat import compat_str
+
+
+class KanaldIE(InfoExtractor):
+    "Kanal D TV Website extractor"
```
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: This line does not make any sense.

File: youtube_dl/extractor/kanald.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    IE_NAME = 'Kanal D'
+    _VALID_URL = r'https?://(?:www\.)?kanald\.com\.tr/(?:.*)/(?P<id>.*\d+.*bolum(?!ler).*)/?'
```
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: `/?` does not make any sense.

File: youtube_dl/extractor/kanald.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            'thumbnail': thumbnail,
+            'description': description,
+            'url': video_url,
+            'uploader': compat_str('Kanal D'),
+            'release_date': release_date
+        }
+
+
+class KanaldSerieIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?kanald\.com\.tr/(?P<id>.*)/(?:bolum|bolumler)$'
```
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: No trailing $, override suitable.

File: youtube_dl/extractor/kanald.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        'https://www.kanald.com.tr/ask-ve-gunah/bolumler/ask-ve-gunah-120-bolum-final',
+        'only_matching': True
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        title = self._html_search_regex(r'<h1>(.+?)</h1>', webpage, 'title')
+        video_url = "https://soledge13.dogannet.tv/" + self._search_regex(
+            r'["\']contentUrl["\']:["\'](?P<video_url>.*)["\']', webpage,
```
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: Do not capture empty strings.

File: youtube_dl/extractor/kanald.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            'ext': 'm3u8',
+            'title': '1.Bölüm',
+            'release_date': '20110907',
+            'thumbnail': r're:^https?://.*\.jpg$',
+            'uploader': 'Kanal D',
+            'description': '1.Bölüm'
+        }
+    }, {
+        'url':
+        'https://www.kanald.com.tr/sevdanin-bahcesi/bolumler/sevdanin-bahcesi-2-bolum',
```
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: Do not carry dict values.

File: youtube_dl/extractor/kanald.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+from re import findall
+
+from .common import InfoExtractor
+from ..compat import compat_str
+
+
+class KanaldIE(InfoExtractor):
+    "Kanal D TV Website extractor"
+    IE_NAME = 'Kanal D'
```
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: Remove.

File: youtube_dl/extractor/kanald.py (hunk 2 above)
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: Must not allow https://www.kanald.com.tr//1-bolum/10115.

File: youtube_dl/extractor/kanald.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        title = self._html_search_regex(r'<h1>(.+?)</h1>', webpage, 'title')
+        video_url = "https://soledge13.dogannet.tv/" + self._search_regex(
+            r'["\']contentUrl["\']:["\'](?P<video_url>.*)["\']', webpage,
+            'video_url')
+        formats = self._extract_m3u8_formats(video_url, video_id)
+        thumbnail = self._search_regex(
+            r'<meta itemprop=["\']thumbnailUrl["\'] content=["\'](?P<thumbnail>.*)["\'].*',
```
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: `.*` at the end does not make any sense.

File: youtube_dl/extractor/kanald.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+        year = self._search_regex(
+            r'["\']uploadDate["\']:["\'](?P<year>\d{4}).*["\']', webpage,
+            'year')
+        month = self._search_regex(
+            r'["\']uploadDate["\']:["\']\d{4}-(?P<month>\d\d).*["\']', webpage,
+            'month')
+        day = self._search_regex(
+            r'["\']uploadDate["\']:["\']\d{4}-\d\d-(?P<day>\d\d).*["\']',
+            webpage, 'day')
+        release_date = year + month + day
```
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: _html_search_meta, unified_timestamp.

File: youtube_dl/extractor/kanald.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+        return {
+            'id': video_id,
+            'title': title,
+            'formats': formats,
+            'thumbnail': thumbnail,
+            'description': description,
+            'url': video_url,
```
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: url and formats are not used together.

File: youtube_dl/extractor/kanald.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+        title = self._html_search_regex(r'<h1>(.+?)</h1>', webpage, 'title')
+        video_url = "https://soledge13.dogannet.tv/" + self._search_regex(
+            r'["\']contentUrl["\']:["\'](?P<video_url>.*)["\']', webpage,
+            'video_url')
+        formats = self._extract_m3u8_formats(video_url, video_id)
+        thumbnail = self._search_regex(
+            r'<meta itemprop=["\']thumbnailUrl["\'] content=["\'](?P<thumbnail>.*)["\'].*',
+            webpage, 'thumbnail')
```
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: Must not be fatal. Read coding conventions.

File: youtube_dl/extractor/kanald.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            page_entries = findall(
+                r'<a.*class=["\']title["\'].*href=["\'](?P<relative>.*)["\'].*',
+                webpage)
+
+            for entry in page_entries:
+                entries.append(
+                    self.url_result('https://www.kanald.com.tr%s' % entry,
+                                    ie=KanaldIE.ie_key()))
+
+        return self.playlist_result(entries, playlist_id)
```
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: generator or PagedList instead of list.

File: youtube_dl/extractor/extractors.py [hunk 12]
```diff
@@ -1524,3 +1524,4 @@
 from .zdf import ZDFIE, ZDFChannelIE
 from .zingmp3 import ZingMp3IE
 from .zype import ZypeIE
+from .kanald import KanaldIE, KanaldSerieIE
```
Comment (dstftw) [inline youtube_dl/extractor/extractors.py]: Alphabetic.

File: youtube_dl/extractor/kanald.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    }, {
+        'url': 'https://www.kanald.com.tr/yarim-elma/bolum/yarim-elma-36-bolum',
+        'only_matching': True
+    }, {
+        'url': 'https://www.kanald.com.tr/ask-ve-gunah/bolumler/ask-ve-gunah-120-bolum-final',
+        'only_matching': True
+    }]
+
+
+class KanaldEmbedIE(KanaldBaseIE):
```
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: Rename to `KanalDIE`.

File: youtube_dl/extractor/kanald.py [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+            json_ld.update({
+                'contentUrl': 'https://soledge13.dogannet.tv/%s' % json_ld['contentUrl']
+            })
+
+        ld_info = self._json_ld(json_ld, video_id)
+
+        return merge_dicts(ld_info, info)
+
+
+class KanaldIE(KanaldBaseIE):
```
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: Rename to something else. This extractor must delegate to embed extractor.

File: youtube_dl/extractor/kanald.py [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+        webpage = self._download_webpage(url, video_id)
+
+        info = {
+            'id': video_id,
+        }
+
+        """FIXME: https://www.kanald.com.tr/kuzeyguney/80-bolum-izle/19364 -> Invalid control character at: line 5 column 146 (char 255)"""
+
+        search_json_ld = self._search_regex(
+            r'(?is)<script[^>]+type=(["\'])application/ld\+json\1[^>]*>(?:\s+)?(?P<json_ld>{[^<]+VideoObject[^<]+})(?:\s+)?</script>', webpage, 'JSON-LD', group='json_ld')
```
Comment (dstftw) [inline youtube_dl/extractor/kanald.py]: `_search_json_ld`.

Comment (mrfade) [reply youtube_dl/extractor/kanald.py]: I think `_search_json_ld` finds the first matching. There are 3 ld+json in the page and `VideoObject` is the last. Also code breaks [in this line](https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/extractor/common.py#L1243) due to `contentUrl` parameter doesn't contains any hostname. It has only path. So must be added hostname to `contentUrl` before calling `_json_ld` function I think

**Pull request closed without merging**